### PR TITLE
Create LOINC Data Files and Validation Positive Pairs

### LINKED ISSUE #104: Create validation data set
With all of our augmentation functions complete, we now need to create a data set on which to run validation. The scope of this ticket is refactoring the augmentation code to allow for the creation of a positive pair validation file.

### PR DESCRIPTION
**Pull request by bamader**
## Description

This PR does several main things:

* Refactors the augmentation code to address a number of bugs; I won't list them all (many of them are in the code comments as new explanations), but we had problems in our character deletion and throughout the enhancement code. That's now fixed and operating better than before (notably because we now jointly model both single-token enhancements and substrings).
* Updates unit tests accordingly
* Generates a set of LOINC data files which can be used for positive pair creation
* Generates a new validation set of positive-pair data which can be run through our performance notebook
* Refactors the performance notebook into performance and embedding so that the functionality of each is more modular and independent

## Related Issues

Closes #104 

## Additional Notes

Apologies for how much got bundled into this PR! I didn't intend to creep it up with so much new code, but between the bugs/refactor and a lot of quality of life stuff...well, here we are 😅

### CODE REVIEW

File: data_curation/augmentation.py [hunk 1]
```diff
@@ -244,110 +257,220 @@ def enhance_loinc_str(
...
+    enhancements_applied = 0
+    enhancements_used = set()
+    num_tries = 0
+
+    # We'll need to generate all enhancements before applying any and then
+    # work backwards by string index; otherwise, we could change single words
+    # into multiples early in the string and ruin the indices of all words
+    # that come later
+    enhancements_to_apply: list[typing.Tuple[list[int], str]] = []
+    while enhancements_applied < num_enhancements and num_tries < MAX_AUGMENTATION_TRIES:
```
[the remark this answers is not in the document]
Comment (bamader) [reply data_curation/augmentation.py]: I did get it to loop a couple times, yeah. The real value-add of the max number of tries is that we're now no longer popping, deleting, or otherwise modifying the list of enhancements that can be chosen. In the old framework, we removed a possible enhancement that could be considered once it was randomly chosen by the function. But now, because we're working with the combined sets, that got uglier to track. I removed the popping/deletion component and instead just tracked our attempts (which is also what we're doing with character deletion so I figured it was a bit more consistent), since every random choice is now from the full pool of available enhancements. In testing I didn't see any instances ever get anywhere close to MAX_TRIES, but it's a nice safeguard to have in place just in-case the random number generator gets really unlucky somehow.

File: data_curation/augmentation.py [hunk 2]
```diff
@@ -244,110 +257,220 @@ def enhance_loinc_str(
...
+        # Base case: singletons are easy to replace
+        if e[0][0] == e[0][1]:
+            words[e[0][0]] = (e[1], e[0])
+        # Substring case involves replacing out a list of tokens
+        # We can just throw away all tokens after the start index and insert
+        # the whole enhancement into one position
+        else:
+            for j in range(e[0][1], e[0][0], -1):
+                del words[j]
+            words[e[0][0]] = (e[1], e[0])
```
[the remark this answers is not in the document]
Comment (bamader) [reply data_curation/augmentation.py]: Oh that's nifty! I didn't know slices could be singularly replaced like that. Yeah I like this more, it looks cleaner on both counts.

File: data_curation/augmentation.py [hunk 3]
```diff
@@ -374,7 +497,7 @@ def generate_augmented_examples(
     """
 
     augmented_examples = []
-    for _ in range(num_examples):
+    for i in range(num_examples):
```
[the remark this answers is not in the document]
Comment (bamader) [reply data_curation/augmentation.py]: Oh whoops, I changed that to an `i` for debugging purposes but forgot to change back.

File: utils/path.py [hunk 4]
```diff
@@ -37,18 +37,32 @@ def read_json(path: str) -> dict:
...
+    matches = glob.glob(pattern)
     enhancements = {}
 
     for match in matches:
-        m = read_json(match)
+        relative_normalized_match = "/".join(match.split("/")[levels:])
+        m = read_json(relative_normalized_match)
         enhancements.update(m)
 
     return enhancements
```
[the remark this answers is not in the document]
Comment (bamader) [reply utils/path.py]: Ah yeah I totally forgot the slash difference would rear up. Good call on using Pathlib to fix.

[the remark this answers is not in the document]
Comment (bamader) [reply utils/path.py]: This code almost worked for me but was still prepending the `..` level prefixes even after the `pathlib.Path(match)` line. I just had to turn that into `match.resolve()` to get Pathlib to turn it into a usable absolute path. See if that works for you as well.

File: data_curation/augmentation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bamader) [reply data_curation/augmentation.py]: Implemented fix

File: utils/path.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (bamader) [reply utils/path.py]: Actually this breaks in github for some reason, need to keep looking at it

File: model_tuning/embedding.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+from typing import List
+
+import torch
+from sentence_transformers import SentenceTransformer
+
+sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))
+
+from utils.parse_and_extract_loinc_names import parse_snoinc_extracts
+
+SNOINC_CODES_FILE = "../data/snoinc_extracts/loinc_lab_names_20251008.csv"
```
Comment (BradySkylight) [inline model_tuning/embedding.py]: I'm fine with hard coding this file name here if we have to, but if we do come out with newer versions of the SNOINC extracts we will have to update this here.

Comment (bamader) [reply model_tuning/embedding.py]: Yeah for sure. With the embedding notebook, we'll only need to run this prior to actual prod deployments, at least for now. We'll be  making all our embeddings before we ever upload to AWS and then dumping them in a vector DB, so this won't need to get touched until we implement functionality to handle changes in LOINC.

File: tests/unit/test_augmentation.py [hunk 6]
```diff
@@ -47,7 +56,7 @@ def test_random_char_deletion(self):
         assert result == expected_result
 
         test_string = self.LOINC_LAB_TEXT_3
-        expected_result = "This term is intnded to collate similar measurements fr he OINC SNOMED CT Collaboration in an ontological view. Addtionally, it can be used to communicate a laboraory order, either alone or in combiaion with specimen or oter information in the odr. It may NOT be sed to report back the measured patient value."
+        expected_result = "Thi term is intnded to collate similar measurements for the LOINC SNOMED CT Collaboration i a ontological view. Addtionally i can be used to communicate a laboraory order, either alone or in combinaion with specimen or other information in the odr. It may NOT be sed to report back the measured patient value."
```
[the remark this answers is not in the document]
Comment (bamader) [reply tests/unit/test_augmentation.py]: Oh maybe? This is showing up for me as a suggested commit change from you, so if that's not intended then def ignore me.

**Comment by bamader** [comment]
Okay all comments addressed, indices removed for readability, tests repaired, path is good to go, and all the good stuff. I think we're ready for a re-review.
